jumanji/environments/routing/pacman/env.py

```diff
@@ -0,0 +1,1012 @@
+            ghost_action: int,
+            ghost_num: int,
+            player_pos: chex.Array,
+            ghost_init_target: chex.Array,
+            old_ghost_locations: chex.Array,
+            ghost_start: chex.Array,
+            scatter_target: chex.Array,
+        ) -> Tuple[chex.Array, int]:
+            """Move a single ghost"""
+            act = jnp.array([0, 1, 2, 3])
+
+            subkey, subsubkey = jax.random.split(subkey)
+            _, valids, valid_actions = self.check_ghost_wall_collisions(
+                state,
+                ghost_pos,
+                ghost_num,
+                player_pos,
+                ghost_init_target,
+                old_ghost_locations,
+                scatter_target,
+            )
+            vert_col = jnp.array([1, 0, 1, 0])
+            hor_col = jnp.array([0, 1, 0, 1])
+
+            # If valids is  [1,0,1,0] or [0,1,0,1] then use old action
+            # this is the case where the ghosts are in a tunnel as ghosts
+            # are not allowed to backtrack.
+            condition = jnp.logical_or(
+                jnp.array_equal(valids, vert_col), jnp.array_equal(valids, hor_col)
+            )
+            condition = jnp.all(condition)
+
+            def is_tunnel(inputs: Tuple[chex.Array, chex.Array, int]) -> int:
+                """Repeat old action if in tunnel"""
+                _, _, ghost_action = inputs
+                return ghost_action
+
+            def no_tunnel(
+                inputs: Tuple[chex.Array, chex.Array, int, chex.PRNGKey, int]
+            ) -> Any:
+                """Chose new action when at intersection"""
+                valids, act, _, key0, _ = inputs
+                key0, subkey0 = jax.random.split(key0)
+                logits = valids
+                logits2 = jnp.where(logits == 0, -1e9, logits)
+                # Softmax distribution
+                softmax_dist = nn.softmax(logits2)
+                a2 = act * logits
+                new_action = jax.random.choice(subkey0, a2, p=softmax_dist).astype(int)
+                return new_action
+
+            inputs_tunnel = (valid_actions, act, ghost_action)
+            inputs_no_tunnel = (valid_actions, act, ghost_action, subsubkey, ghost_num)
+
+            result_true = is_tunnel(inputs_tunnel)
+            result_false = no_tunnel(inputs_no_tunnel)
+
+            def start_over() -> Any:
+                """If not in waiting mode then pick new action"""
+                chosen_action = jnp.where(
+                    condition[..., None], result_true, result_false
+                )
+                chosen_action = jnp.squeeze(chosen_action)
+                return chosen_action
+
+            def no_start() -> int:
+                """If in waiting mode then use no-op"""
+                chosen_action = 4
+                return chosen_action
+
+            position = ghost_pos
+            chosen_action = jax.lax.cond(ghost_start < 0, start_over, no_start)
```

Comment:
Instead of computing both branches rather pass the args to `cond`.

Make `start_over` and `no_start` take `valid_actions, act, ghost_action, subsubkey, ghost_num` as args and call `is_tunnel`/`no_tunnel` inside of `start over` then pass those args to the `cond`.

Reply:
Updated as suggest. It is a lot cleaner.